### DISCUSSION TITLE
Plugin update

### DISCUSSION THREAD

**Discussion by adorobis** [tprelog/iocage-homeassistant#54] [Q&A]
@tprelog  as suggested let's have a separate discussion on the plugin update.
So you are saying that running the update plugin:
![image](https://user-images.githubusercontent.com/19752098/152401596-f10b4efb-4056-4625-ad2f-0413bb3d2273.png)
even without checking the "Update jail as well" option will update something more than just the plugin overlay files?

**Comment by tprelog** [comment]
That's correct. Though, I'm not exactly sure what "Update jail as well" does in addition vs leaving unchecked

Here's my understanding about the Plugin UPDATE process

1. It takes a snapshot
2. Clones the plugin repository
3. Applies the [overlay folder ](https://github.com/tprelog/iocage-homeassistant/tree/master/overlay)to the plugin root fs
4. Run the `pre_update.sh` script, if it exists
5. Remove all user land installed packages
6. Install up to date versions of all packages listed in the [plugin manifest](https://github.com/ix-plugin-hub/iocage-plugin-index/blob/master/homeassistant.json)
7. Run the `post_update.sh` script, if it exists

I know I've changed a few things since your version of the plugin but I think the overlay is still compatible, with a few changes. There's nothing that should affect Home Assistant (unless your plugin is over 2.5 years old) but for example the `menu` script location has changed. We would need to fix-up a few things so the new menu started when you enter the jail console

If you run the Plugin UPDATE I try to do all that in the `pre_update.sh` script. Actually, I think from your version it will try to just reinstall Home Assistant automatically. I know recreating the virtualenv was a hard requirement when version 5 came out. With version 5 I had to cut ties with all the previous crap I was trying to do. I really wanted everyone to do a clean install at that point but troubleshooting a failed Plugin UPDATE is still feasible, with a little time and patience of course.

The risk when Home Assistant, or the jail itself seems to break is usually when the system packages get updated.

When there is an issue with Home Assistant after such an update, the solution is usually to just recreate the virtualenv.

```
service homeassistant reinstall homeassistant
```

Do you want to try a Plugin UPDATE. Since you still have the 12.1-RELEASE, I think I would try the "Update Jail as well" to see if you get to the current 12.2-RELEASE-p10. Or we might need to try some `iocage` commands to achieve that RELEASE upgrade

We can also try to just do the overlay manually, if you want to try that.

Or if you're using an external dataset for your config the quickest way forward is honestly to just install a new plugin and mount your existing config to it... 

I will try to help with any path forward you choose

**Reply by adorobis** [reply]
Many thanks for quick answer. A few options to consider :) The quickest would by just manual update of plugin overlay files, the most safe probably the clean plugin install (I have ha in external dataset). 

> Do you want to try a Plugin UPDATE.

I have bad experience with Plugin UPDATE for some other plugins. It quite often does not complete successfully.

I have to think about a few additional steps e.g. IP address, additional service I have in the jail - probably would be safer to move it to a dedicated jail - probably I need to start with this before installing clean plugin. The good thing is that I have vast majority of other components in dedicated jails (mosquito, zwave, zigbee, mariadb etc).

Will let report back which path I choose and hopefully no additional support will be needed. But many thanks for offer!

**Reply by tprelog** [reply]
Sounds good!

All the convenience of this plugin should work just fine in a regular jail as well. I know I'm terrible at documentation but there's no magic here. I have peoples general distaste (including my own, it's a love / hate thing, really) for plugins in mind from the beginning.

**Reply by adorobis** [reply]
So I've found a bit of time to go through it. Installed HA plugin from scratch and mounted the configuration directory. And all works right away. The only thing was that it was a bit slow start after mounting the config directory.

Now I have `plugin version v7-2022.2.3` and jail release `12.2-RELEASE-p11`. Btw when creating empty jail a release `12.3-RELEASE` is used - is it something you control in the plugin?
The only manual thing I have to do is to copy some scripts over to the new jail (used in command_line sensors). Have to move it to the config directory so no manual intervention would be required.

**Reply by adorobis** [reply]
Ah, and btw, again the pymysql module was not installed automatically and I've had to add it manually to the venv.

**Reply by tprelog** [reply]
> again the pymysql module was not installed automatically and I've had to add it manually to the venv.

Yea, in the venv, if it's something extra you installed manually to begin with, it will always need re-installed manually as well. When it comes to reinstalling the virtualenv within an existing jail, there is a pip command (I'd have to look it up) that can be used to generate a list of everything installed in the venv. This list can then be passed as requirements to the install command.

```
service homeassistant install homeassistant -r /path/to/file
``` 

> The only thing was that it was a bit slow start after mounting the config directory.

This is normal... What's happening behind the scenes is that Home Assistant is reinstalling all the integrations you have. This can take a while since lots of things may need to be compiled (aka building wheels). There's actually a trick you can use to watch these start-ups.

As you may know, the log level in Home Assistant is set to `warning` unless otherwise configured in `configuration.yaml`. For this reason nothing shows up until something is wrong. Basically, no news is good news. But you want to know the news regardless... So for this you need to set Home Assistant's [logger integration](https://www.home-assistant.io/integrations/logger/) to `info`

```yaml
logger:
  default: info
```

You can also use this next command to view and "follow" the log. This can be useful to have open in a second window during an upgrade or when reinstall the virtualenv, like when python needs updated.

```
service homeassistant logs -f
```

I made a quick video (no audio) showing how this works using a service restart as an example. 

Here I am logged into the jail twice. On the left I'm using the menu to restart the service. On the right, I have exited the menu and entered the command to view and "follow" the logfile. During the first restart I am just using the default configuration. On the right side of the screen you no additional information is printed.

Then I quickly edit the `homeassistant/configuration.yaml` to set the logger level to `info` and re-run the command to "follow" the logs.

When I restart Home Assistant a second time, you can see all the additional information about what's starting.

https://user-images.githubusercontent.com/35702532/152890369-0d258c97-6f78-4900-b3de-c6863302ba86.mp4

BTW there's a daemon log: `/var/log/homeassistant_daemon.log` -- Basically it the same as `home-assistant.log`

```
service homeassistant logs -f
```

Is a shortcut for
```
tail -F /var/log/homeassistant_daemon.log
```

[I threw a few others options  in as well](https://github.com/tprelog/iocage-homeassistant/blob/8f454586c077243f19621181125166c3060e66fe/overlay/usr/local/etc/rc.d/homeassistant#L210)

---

> Btw when creating empty jail a release `12.3-RELEASE` is used - is it something you control in the plugin?

Sort of... maybe... but definitely not from the usual ix-system plugin repo.  I'm not well versed in FreeBSD to understand any implications of this, but since the TrueNAS host is still running 12.2-RELEASE it is only that userland application that will be from 12.3-RELEASE. I can not say what effect, if any, this will have when Home Assistant is building the wheels.

I was running some tests on different jail types, trying to figure out options for user who do not upgrade to TrueNAS CORE 13 immediately upon release. This is when I noticed Home Assistant still appears to build wheels for the 12.2-RELEASE, even though the jail was created as 12.3 -- https://github.com/ix-plugin-hub/iocage-plugin-index/issues/284#issuecomment-1003105752

At this time ix-systems has no plans to bump plugins to the 12.3-RELEASE. The next release will be for 13.0, but I have a [separate plugin repo containing only the  plugin manifest I maintain](https://github.com/tprelog/truenas-plugin-index). I could try to publish a 12.3-RELEASE branch there and then try switching the repository on a plugin.

```
iocage set  plugin_repository=https://github.com/tprelog/truenas-plugin-index.git
```

From there try to somehow upgrade to the desired release.  I'm not sure if, or how that might be possible. Definitely need to do some testing to figure out. I think we need to use `iocage upgrade` to change the release. I think, it should  be the same as upgrading a regular jail

**Reply by tprelog** [reply]
What the heck, I just went and recorded in real time - reinstalling the Home Assistant virtualenv with HA open in the browser and a window "following" the log file... It's long.... It's boring but I think it gives a good idea of what to expect. With the logger still set to `info`, you see the file come alive as soon as the installation finishes. Around 3 min 10 sec we see home assistant is now listening on port 8123. About 10 seconds later you may notice the toast banner in home assistant changes from connection lost to Home Assistant is starting - this is followed a pause in activity... I know something's happening because I hear the fans on my computer kick up. Finally the logs start to roll again... It is not until around 6 min 50 secs that we finally see Home Assistant has started and it is still continuing to install additional modules for things as they are being discovered on my network.

https://user-images.githubusercontent.com/35702532/152909688-a6a6cfdf-e2cc-4a23-aec4-e5b4cc339c37.mp4

**Reply by adorobis** [reply]
Wow that's a comprehensive response! Very useful for future troubleshooting. As the fresh install went ok, I'm not going to dig deeper now.
My only comment would be regarding the pymysql package: I have not installed it manually originally as far as I remember, I only need or for recorder supporting data to mariadb, which is a standard feature. Shouldn't HA install this package automatically?
Regarding TrueNAS 13, let me know if you'll need any testing to make your plugin work on it, I'll try to set up a test system.
